### PR TITLE
New `heading` command

### PR DESCRIPTION
**Pull request by byorgey**
Define a new command `heading : cmd dir` which returns the robot's current heading (or `down` if it doesn't have one, or if the heading is some weird vector).

My immediate motivation is that I want to be able to define `excursion : cmd unit -> cmd unit` which executes the given command and then returns to the same location and orientation as before.  The `heading` command is one of the last missing pieces of machinery necessary to be able to implement this.

### CODE REVIEW

**Review by kostmo — APPROVED** [review]
Makes sense.

File: src/Swarm/Language/Syntax.hs [hunk 1]
```diff
@@ -104,6 +107,9 @@ instance ToJSONKey AbsoluteDir where
 instance FromJSONKey AbsoluteDir where
   fromJSONKey = genericFromJSONKey defaultJSONKeyOptions
 
+-- | A relative direction is one which is defined with respect to the
```
Comment (xsebek) [inline src/Swarm/Language/Syntax.hs:110]: I like this definition, it makes sense to have `DDown` relative this way. 👍

File: src/Swarm/Game/Step.hs [hunk 2]
```diff
@@ -1003,6 +1003,17 @@ execConst c vs s k = do
...
+      mh <- use robotOrientation
+      -- In general, (1) entities might not have an orientation, and
+      -- (2) even if they do, orientation is a general vector, which
+      -- might not correspond to a cardinal direction.  We could make
+      -- 'heading' return a 'maybe dir' i.e. 'unit + dir', or return a
+      -- vector of type 'int * int', but those would both be annoying
+      -- for players in the vast majority of cases.  We rather choose
+      -- to just return the direction 'down' in any case where we don't
+      -- otherwise have anything reasonable to return.
+      return $ Out (VDir (fromMaybe (DRelative DDown) $ mh >>= toDirection)) s k
```
Comment (xsebek) [inline src/Swarm/Game/Step.hs:1016]: Why not use `maybe`?

**Review by xsebek — APPROVED** [review]
LGTM! 👍

File: src/Swarm/Language/Syntax.hs (hunk 1 above)
Comment (byorgey) [reply src/Swarm/Language/Syntax.hs:110]: Yes, I was confused about `DDown` so I added these comments after I figured it out. :smile:

File: src/Swarm/Game/Step.hs (hunk 2 above)
Comment (byorgey) [reply src/Swarm/Game/Step.hs:1016]: How would you use it, exactly?

Comment (xsebek) [reply src/Swarm/Game/Step.hs:1016]: Never mind, I thought this was `maybe (DRelative Down) toDirection mh` but I see now its two `Maybe`s. 👍

**Pull request merged**